### ISSUE TITLE
Reading Multiple Blocks

### ISSUE THREAD

**Issue by iamlemec** [alexcrichton/bzip2-rs#41]
I'm encountering a slight issue when using the library: I'm only getting the first block of bz2 files. In my case, the block size is 900k, and when `read`ing off `BzDecoder`, I get a total of 900k then EOF (size zero reads thereafter).

This happens on both the crates.io release and master here on github. Am I confused here? Would appreciate any suggestions. Thanks!

**Comment by alexcrichton** [comment]
This seems like it's likely a bug in this library, but unfortunately I wouldn't know where to start. Do you have sample files that I can help poke around at?

**Comment by iamlemec** [comment]
Ok, new clue. When preparing the example file, I realized that I had compressed my files with pbzip2 (parallel bzip2). If I compress things with regular bzip2, everything works fine and bzip2-rs yields the entire file. I wonder if it's that pbzip2 is using a different blocking scheme or some unusual options? The pbzip2 files decompress as expected with regular bunzip2.

This is probably kind of an edge case, so I would understand if it doesn't have high priority, but here is an sample file compressed with pbzip2: https://send.firefox.com/download/51320afe37/#ECAJud4iGH-7UMjJ7M8Pjw

**Comment by alexcrichton** [comment]
Ah an interesting observation! I wonder if this has to do with a format that the bunzip2 tool specifically allows?

One thing we ran into with `flate2-rs` is that you can literally concatenate two compressed gzip files to create a new one, and then when decoding you'll decode both back-to-back. I wonder if that's what's happening here? Is pbzip2 creating concatenated compressed streams?

**Comment by iamlemec** [comment]
Yup, looks like it! Just grepping through the file, I'm seeing multiple BZ headers (which, appropriately for today, is BZh9 + 0x314159265359). If I change the logic in BzDecoder::read to only stop on EOF and keep going after StreamEnd, it works like a charm.

**Comment by iamlemec** [comment]
Spoke too soon there. You need to call bzlib restart too. Unfortunately, this results in a large memory leak for me, and I know almost nothing about rust memory management, especially with ffi. Here's what I have so far: https://github.com/iamlemec/bzip2-rs/commit/c0c501d6c6fff648fc31b9a3f71ff55f667755a4

**Comment by alexcrichton** [comment]
@iamlemec oh I think you can solve that by replacing `self.data.restart(false)` with `self.data = Decompress::new(false)`

**Comment by iamlemec** [comment]
Works great now. Thanks!

**Comment by alexcrichton** [comment]
@iamlemec were you thinking of sending a PR akin to `MultiGzDecoder` for this crate perhaps?

**Comment by iamlemec** [comment]
Definitely. I'll check out `MultiGzDecoder` and submit something analogous relatively soon.

**Comment by andreycizov** [comment]
Thanks @iamlemec for the initial work!

I have to admit though that I wasn't happy with the results, as the decompression speed at one thread sets me at 20MB/s so it's still faster to transcode to gzip with `pbzip2` first and then process the results with rust.

https://github.com/alexcrichton/bzip2-rs/pull/44
